contracts/epochs/FlowEpoch.cdc

```diff
@@ -109,46 +121,511 @@ pub contract FlowEpochs {
             self.collectorClusters = collectorClusters
             self.clusterQCs = clusterQCs
-            self.dkgGroupKey = dkgGroupKey
+            self.dkgKeys = dkgKeys
+        }
+
+        pub fun setClusterQCs(qcs: [FlowEpochClusterQC.ClusterQC]) {
+            self.clusterQCs = qcs
+        }
+
+        pub fun setDKGGroupKey(keys: [String]) {
+            self.dkgKeys = keys
+        }
+    }
+
+    /// Metadata that is managed and can be changed by the Admin///
+    pub struct Config {
+        /// The number of views in an entire epoch
+        pub(set) var numViewsInEpoch: UInt64
+
+        /// The number of views in the staking auction
+        pub(set) var numViewsInStakingAuction: UInt64
+        
+        /// The number of views in each dkg phase
+        pub(set) var numViewsInDKGPhase: UInt64
+
+        /// The number of collector clusters in each epoch
+        pub(set) var numCollectorClusters: UInt16
+
+        init(numViewsInEpoch: UInt64, numViewsInStakingAuction: UInt64, numViewsInDKGPhase: UInt64, numCollectorClusters: UInt16) {
+            self.numViewsInEpoch = numViewsInEpoch
+            self.numViewsInStakingAuction = numViewsInStakingAuction
+            self.numViewsInDKGPhase = numViewsInDKGPhase
+            self.numCollectorClusters = numCollectorClusters
+        }
+    }
+
+    access(contract) let configurableMetadata: Config
+
+    /// Metadata that is managed by the smart contract
+    /// and cannot be changed by the Admin
+
+    /// The counter, or ID, of the current epoch
+    pub var currentEpochCounter: UInt64
+
+    /// The counter (ID) for the next epoch
+    pub var proposedEpochCounter: UInt64
+
+    /// The current phase that the epoch is in
+    pub var currentEpochPhase: EpochPhase
+
+    /// Contains a historical record of the metadata from all previous epochs
+    /// indexed by epoch number
+    access(contract) var epochMetadata: {UInt64: EpochMetadata}
+
+    access(contract) let QCAdmin: @FlowEpochClusterQC.Admin
+
+    access(contract) let DKGAdmin: @FlowDKG.Admin
+
+    access(contract) let stakingAdmin: @FlowIDTableStaking.Admin
+
+    pub let adminStoragePath: StoragePath
+    pub let heartbeatStoragePath: StoragePath
+
+    /// Resource that can update some of the contract fields
+    pub resource Admin {
+        pub fun updateEpochViews(_ newEpochViews: UInt64) {
+            pre {
+                FlowEpoch.currentEpochPhase == EpochPhase.STAKINGAUCTION: "Can only update fields during the staking auction"
+                newEpochViews > FlowEpoch.configurableMetadata.numViewsInStakingAuction: "Epoch Views must be greater than staking views"
+                newEpochViews > FlowEpoch.configurableMetadata.numViewsInDKGPhase: "Epoch Views must be greater than DKG phase views"
+            }
+
+            FlowEpoch.configurableMetadata.numViewsInEpoch = newEpochViews
+        }
+
+        pub fun updateAuctionViews(_ newAuctionViews: UInt64) {
```

Comment:
I really like the way you have structured these methods, It makes it easy to reason about the state machine.